### PR TITLE
Rewritten mods section

### PR DESCRIPTION
**Pull request by knoxfighter**
Finally ready, i rewrote the whole Mods Section, so it is possible to easily download and update mods via the mod-portal-api and managing modpacks.

### CODE REVIEW

**Comment by mroote** [comment]
Awesome! I'll hopefully have some time soon to test this out.

To fix the error in Travis you will likely have to add any new dependencies to the travis.yml file in the root of the repo.

File: src/lockfile/lockfile.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
+package lockfile
```
Comment (mroote) [inline src/lockfile/lockfile.go]: Move this package to a new vendor directory and then you don't need to change the $GOPATH.

```
src/
├── vendor
│   └── lockfile
│       └── lockfile.go

```

File: src/mods_test.go [hunk 2]
```diff
@@ -1,37 +1,46 @@
 package main
```
Comment (mroote) [inline src/mods_test.go]: I think we can just remove this file if all the tests are commented out. Unless you want to add a few tests!

File: src/routes.go [hunk 3]
```diff
@@ -152,32 +152,67 @@ var apiRoutes = Routes{
...
+		"POST",
+		"/mods/delete/all",
+		DeleteAllModsHandler,
 	}, {
-		"DownloadMod",
+        "UpdateMod",
+        "POST",
+        "/mods/update",
+        UpdateModHandler,
+    }, {
```
Comment (mroote) [inline src/routes.go]: Formatting

File: ui/App/components/Mods/Mod.jsx [hunk 4]
```diff
@@ -1,33 +1,143 @@
...
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if(prevProps.mod.version != this.props.mod.version) {
+            this.checkForNewVersion();
+        }
+    }
+
+    checkForNewVersion() {
+        let this_class = this;
```
Comment (mroote) [inline ui/App/components/Mods/Mod.jsx]: You don't need to change this with ES6 Javascript.

In the Component class you can access it's props directly like `this.props.updateCountAdd`

File: ui/App/components/ModsContent.jsx [hunk 5]
```diff
@@ -1,121 +1,572 @@
...
+        this.updateAllModsHandler = this.updateAllModsHandler.bind(this);
+        this.updatesAvailable = this.updatesAvailable.bind(this);
+        this.updateCountSubtract = this.updateCountSubtract.bind(this);
+        this.updateCountAdd = this.updateCountAdd.bind(this);
+
         this.state = {
-            installedMods: [],
-            listMods: [],
-            modPacks: [],
+            logged_in: false,
```
Comment (mroote) [inline ui/App/components/ModsContent.jsx]: Use camelcase here to keep it consistent.

File: ui/App/components/Mods/ModUpload.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+
+class ModUpload extends React.Component {
+    componentDidMount() {
+        $("#mod_upload_input").fileinput({
+            uploadUrl: '/api/mods/upload',
+            // hiddenThumbnailContent: true,
+            // showPreview: false,
```
Comment (mroote) [inline ui/App/components/Mods/ModUpload.jsx]: Remove the commented out code.

File: src/mod_modInfo.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    Destination string `json:"-"`
+}
+type ModInfo struct {
+    Name string `json:"name"`
+    Version string `json:"version"`
+    Title string `json:"title"`
+    Author string `json:"author"`
+    FileName string `json:"file_name"`
+    FactorioVersion string `json:"factorio_version"`
+}
```
Comment (mroote) [inline src/mod_modInfo.go]: Run gofmt on these files to format everything properly.

File: src/mod_Mods.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+func (mods *Mods) updateMod(mod_name string, url string, filename string) error {
+    var err error
+
+    //err = mods.deleteMod(mod_name)
+    //if err != nil {
+    //    log.Printf("updateMod ... error when deleting mod: %s", err)
+    //    return err
+    //}
```
Comment (mroote) [inline src/mod_Mods.go]: Commented out code, just remove it if not needed.

File: ui/App/components/ModsContent.jsx (hunk 5 above)
Comment (knoxfighter) [reply ui/App/components/ModsContent.jsx]: I will change to camelcase everywhere :)

File: src/lockfile/lockfile.go (hunk 1 above)
Comment (knoxfighter) [reply src/lockfile/lockfile.go]: That doesnt work so easy. It seems like the vendor folder needs the additional program govendor to be installed. When i understand it correctly, the vendor folder is an alternative to the normal $GOPATH. Using it with local packages is possible, but i dont think it is so good to do.
I also can upload it to a new github repo and import it from there, if you want.

I would say, keeping it, as it is currently is, will be the easiest way, but i can set up, other things, if you want :)

Comment (mroote) [reply src/lockfile/lockfile.go]: Vendor directory is standard as of Go 1.6, so dependencies should be placed in there and they will work with the standard go commands like `go build`.

When using the go tools such as go build or go run, they first check to see if the dependencies are located in ./vendor/. If so, use it. If not, revert to the $GOPATH/src/ directory.

Comment (knoxfighter) [reply src/lockfile/lockfile.go]: On my environment with Go 1.9.1 on ubuntu 16.04 it cant compile anymore, cause the dependensies cant be found:
```
manager/src$ go build
mod_Mods.go:10:2: cannot find package "lockfile" in any of:
        /usr/local/go/src/lockfile (from $GOROOT)
        /home/knox/Desktop/Programme/GO/factorio-server-manager/manager/src/lockfile (from $GOPATH)
        /home/knox/go/src/lockfile
```

Do i need to add a vendor,json to make it work, or does it work without?

Comment (mroote) [reply src/lockfile/lockfile.go]: No it should just work with the file structure I posted.  I tested on my side and I can build it fine with the vendor directory and I don't have any extra dependency management tools installed.

My src/ directory looks like this:

```
.
├── auth.go
├── factorio_config.go
├── factorio_server.go
├── factorio_server_linux.go
├── factorio_server_windows.go
├── gamelog.go
├── handlers.go
├── main.go
├── mod_modInfo.go
├── mod_modpacks.go
├── mod_Mods.go
├── mod_modSimple.go
├── mods.go
├── mods_handler.go
├── mods_test.go
├── rcon.go
├── routes.go
├── saves.go
├── vendor
│   └── lockfile
│       └── lockfile.go
├── wsclient.go
└── wsroutes.go
```

Comment (knoxfighter) [reply src/lockfile/lockfile.go]: I misunderstood the GOPATH and read more things about it. It is working in my environment too and i updated the fle.

File: ui/App/components/ModsContent.jsx (hunk 5 above)
Comment (knoxfighter) [reply ui/App/components/ModsContent.jsx]: Do you want to have all ReactJS code in camelCase?

File: ui/App/components/Mods/Mod.jsx (hunk 4 above)
Comment (knoxfighter) [reply ui/App/components/Mods/Mod.jsx]: I discovered, that using function() { ... } will override the this-variable inside the function.
Using the lambda-expression () => { ... } for functions, dont override the this-varaible.
... JavaScript is funny xD

File: ui/App/components/ModsContent.jsx (hunk 5 above)
Comment (mroote) [reply ui/App/components/ModsContent.jsx]: Yeah let's make it camelcase here as well.

File: ui/App/components/Mods/Mod.jsx (hunk 4 above)
Comment (mroote) [reply ui/App/components/Mods/Mod.jsx]: Ah I see that's good to know, ES6 added a bunch of nice features to Javascript.  It's still Javascript though.. XD

File: src/lockfile/lockfile.go (hunk 1 above)
Comment (mroote) [reply src/lockfile/lockfile.go]: Nice!

File: src/mod_modInfo.go [hunk 9]
```diff
@@ -111,7 +111,7 @@ func (modInfoList *ModInfoList) deleteMod(modName string) error {
 		}
 	}
 
-	log.Printf("the mod-file doesnt exists!")
+	log.Printf("the mod-file for mod %s doesntt exists!", modName)
```
Comment (mroote) [inline src/mod_modInfo.go]: Small typo here

**Comment by mroote** [comment]
This is looking pretty good thanks for working on this feature!

I made a build with your branch but I'm getting some errors in the frontend. When I try and click upload it doesn't seem to upload the file due to an undefined error. Overall I like this though mod packs have been a long requested feature and this is shaping up great.

screenshot from 2018-01-10 19-27-42

**Comment by knoxfighter** [comment]
I think you made a mistake on building the ui, cause the sreencshot shows the old ui and not my rewritten one. To build the ui you simply has to use `npm install` to get the two added dependencies and then run `npm run build`. When you did this and it doesn't throw any error messages, i think it is a problem with the cache and you get the old page used on the other pull requests :)
The design of the new mods section can be seen on the screensshots i added to the readme.

**Comment by mroote** [comment]
Ah you are correct I was still using the old frontend code :P

This looks great! There are two small conflicts to resolve and then I think we are good to merge

**Comment by mroote** [comment]
Thanks for working on this feature! It really improves the mod section.

**Pull request merged**